File: leetcode/19-remove-nth-node-from-end-of-list/main.go
/*
Problem: Remove Nth Node From End of List
Link: https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
*/
package main

type ListNode struct {
	Val int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{0, head}
  slow, fast := dummy, dummy

  // Move fast to n+1 pos
  for i := 0; i < n+1; i++ {
    fast = fast.Next
  }

  // Move both slow and fast step by step (until fast to the end)
  for fast != nil {
    slow = slow.Next
    fast = fast.Next
  }

  // Let the next of prev of n point to the next of n
  slow.Next = slow.Next.Next

  return dummy.Next
}

File: leetcode/19-remove-nth-node-from-end-of-list/Main.kt
package main

class Solution {
    fun removeNthFromEnd(
        head: ListNode?,
        n: Int,
    ): ListNode? {
        var slow = head
        var fast = head

        // Initialize the fast pointer
        for (i in 1..n) {
            fast = fast?.next
        }
        if (fast == null) return head?.next

        while (fast?.next != null) {
            slow = slow?.next
            fast = fast?.next
        }

        slow?.next = slow?.next?.next

        return head
    }
}
